feat(ring_theory): Fractional ideals

Vierkantor opened the pull request:
This PR defines fractional ideals of an integral domain and proves basic facts about them: they have a `comm_semiring` structure, `lattice` structure and ideal quotient; the inverse of an ideal, if it exists, is equal to the quotient. These is the work in progress I mentioned [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/class.20groups.20of.20number.20fields.3F/near/188214061).

The things that are called "fractional ideal of `R`" in algebraic number theory are called `fractional_ideal R (non_zero_divisors R)` in this file, a notation that probably deserves to be shorter. Do you happen to know a good abbreviation? Alternatively: Is there a good name `foo` for `fractional_ideal` so that we can abbreviate `fractional_ideal R := foo R (non_zero_divisors R)`?

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

Vierkantor committed:
Some WIP work on fractional ideals

Vierkantor committed:
Fill in the `sorry`

Vierkantor committed:
A lot of instances for fractional_ideal

Vierkantor committed:
Show that an invertible fractional ideal `I` has inverse `1 : I`

Vierkantor committed:
Cleanup and documentation

Vierkantor committed:
Move `has_div submodule` to algebra_operations

Vierkantor committed:
More cleanup and documentation

Vierkantor committed:
Explain the `non_zero_divisors R` in the `quotient` section

kim-em commented on src/ring_theory/algebra_operations.lean:
```diff
@@ -222,6 +222,23 @@ begin
+carrier := { x | ∀ y ∈ J, x * y ∈ I },
+zero := λ y hy, by { rw zero_mul, apply submodule.zero },
+add := λ a b ha hb y hy, by { rw add_mul, exact submodule.add _ (ha _ hy) (hb _ hy) },
+smul := λ r x hx y hy, by { rw algebra.smul_mul_assoc, exact submodule.smul _ _ (hx _ hy) } } ⟩
```
```suggestion
carrier := { x | ∀ y ∈ J, x * y ∈ I },
zero    := λ y hy, by { rw zero_mul, apply submodule.zero },
add     := λ a b ha hb y hy, by { rw add_mul, exact submodule.add _ (ha _ hy) (hb _ hy) },
smul    := λ r x hx y hy, by { rw algebra.smul_mul_assoc, exact submodule.smul _ _ (hx _ hy) } } ⟩
```

jcommelin commented on src/ring_theory/algebra_operations.lean:
```diff
@@ -222,6 +222,23 @@ begin
+-/
+instance : has_div (submodule R A) :=
+⟨ λ I J, {
+carrier := { x | ∀ y ∈ J, x * y ∈ I },
```
If you want, you can use `pointwise_mul` for this definition. Then you would be able to write `x \bullet J \sub I`. But that probably has slightly worse definitional properties.

Vierkantor replied:
It is less convenient to work with, so I decided not to change the definition. Instead I added a `mem_div_iff_smul_subset` lemma that converts between the two.

jcommelin commented on src/ring_theory/fractional_ideal.lean:
```diff
@@ -0,0 +1,417 @@
+variables (R : Type u) [integral_domain R] (S : set R) [is_submonoid S]
+
+/-- A submodule `I` is a fractional ideal if `a I ⊆ R` for some `a ≠ 0`. -/
+def is_fractional (I : submodule R (localization R S)) :=
```
When it comes to applying this theory, this definition would benefit enormously from using a characteristic predicate for localisations. For example, it will now be hard (i.e. impossible) to talk about fractional ideals in `rat` wrt to `int`. Because `rat` is not of the form `localization int _`.
@101damnations is working on bringing this characteristic predicate to mathlib. I hope that these two PR will play well together.

Vierkantor replied:
I expect that changing `localization ℤ (non_zero_divisors ℤ)` to, let's say a typeclass `has_localization ℤ ℚ`, shouldn't have too much impact on this PR, since most of it is only concerned with `is_integer` and its lemmas.

jcommelin commented on src/ring_theory/fractional_ideal.lean:
```diff
@@ -0,0 +1,417 @@
+
+This section defines the ideal quotient of fractional ideals.
+
+In this section we need that each non-zero `y : R` has an inverse in
```
Up till now, have you used that `R` is an integral domain? Or would `comm_ring` suffice? If so, I think it would make sense to only start assuming `R` is an integral domain from now on.

Vierkantor replied:
The reason I chose `integral_domain` over `comm_ring` throughout the file is to ensure the definition of `I + J` and `I * J` works: if the denominators for `I` and `J` are `aI ≠ 0` and `aJ ≠ 0` respectively, then `I + J` and `I * J` have denominator `aI * aJ ≠ 0`. (And even Bourbaki only considers fractional ideals as submodules of the fraction field of an integral domain.)

I also considered taking `R` to be a `comm_ring` and have `a ∈ S` (where `S` is the submonoid at which we localize). That didn't work out when I started, but now after I've changed around some things it might work. Does that seem like the appropriate generalization?

jcommelin replied:
My only reason for weakening the assumption would be to make the life of the tc inference system easier. I guess in all applications `R` will be an integral domain. Giving the user an `a ∈ S` proof obligation doesn't sound ideal. So I would prefer to avoid that.

Vierkantor replied:
OK, then I will keep it an integral domain and leave the generalization to `comm_ring` to someone else if they actually need it.

jcommelin commented on src/ring_theory/fractional_ideal.lean:
```diff
@@ -0,0 +1,417 @@
+    simp [y_eq_y'.symm] }
+end
+
+/-- `1 / I` is the inverse of `I` if `I` has an inverse. -/
```
Doesn't this mean that you can now add a `has_inv` instance?

jcommelin commented on src/ring_theory/localization.lean:
```diff
@@ -595,4 +595,85 @@ def le_order_embedding :
+
+/-- `a : localization α S` is an integer if it is an element of the original ring `α` -/
+def is_integer (S : set α) [is_submonoid S] (a : localization α S) : Prop :=
+a ∈ (set.univ : set α).image (coe : α → localization α S)
```
I think that `range` is supposed to be used instead of `(set.univ ...).image`.

jcommelin reviewed (commented):
Nice work! Thanks for this PR!

Vierkantor committed:
whitespace

Co-Authored-By: Scott Morrison <scott@tqft.net>

Vierkantor committed:
`has_inv` instance for `fractional_ideal`

Vierkantor committed:
`set.univ.image` -> `set.range`

Vierkantor committed:
Fix: `mem_div_iff.mpr` should be `mem_div_iff.mp`
(but both reduce to reflexivity anyway)

Vierkantor committed:
Add `mem_div_iff_smul_subset`

Since that is how the definition of `I / J` is traditionally done,
but it is not as convenient to work with, I didn't change the definition
but added a lemma stating the two are equivalent

Vierkantor committed:
whitespace again

(it got broken because I merged changes incorrectly)

Vierkantor committed:
Fix unused argument to `inv_nonzero`

jcommelin reviewed (approved):
LGTM

mergify[bot] committed:
Merge branch 'master' into fractional_ideal
